Oauth

irenenikk opened the pull request:
Add oauth support
Add organizations (new command and prompt on login)

Commit:
Add oauth support, allow aborting download

Commit:
Fix tests

Commit:
Add tests, fix one liner command when slug invalid

nygrenh commented on pom.xml:
```diff
@@ -181,6 +182,9 @@
                 <artifactId>cobertura-maven-plugin</artifactId>
                 <version>2.7</version>
                 <configuration>
+                    <check>
```
Is this necessary?

irenenikk replied:
According to IntelliJ, yes. Without the check-tag, the configuration tag is shown in red.

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/Application.java:
```diff
@@ -20,12 +20,7 @@
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
I'd prefer not to use star imports.

irenenikk replied:
:palm_tree:

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/Application.java:
```diff
@@ -176,6 +172,20 @@ public void run(String[] args) {
     }
 
     public static void main(String[] args) {
+        while(true) {
```
Do we want to have your debugging loop as the main entry point for all of our users?

irenenikk replied:
:dolphin:

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/Account.java:
```diff
@@ -1,35 +1,53 @@
 public class Account {
 
-    static Account NULL_ACCOUNT = new Account(null, null, null);
+    static Account NULL_ACCOUNT = new Account(null, null, null, null);
```
Null null null null null null null null null null null... :camel:

irenenikk replied:
The Null Account is a part of the original software. It is used when the method `setAccount(Account a)` of `Settings` is called with `a` being null, so that the getters wouldn't throw the `NullPointerException`. `setAccount` was added by the original team, and I did not find any use of it where `account` would seem to be null, so maybe deleting it would indeed be a good idea.

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java:
```diff
@@ -49,32 +51,40 @@ public void setWorkDir(WorkDir workDir) {
+        if (account.getServerAddress().isPresent()) {
+            return account.getServerAddress().get();
+        }
+        return "https://tmc.mooc.fi";
```
Maybe we should extract this into a constant?

irenenikk replied:
:tulip:

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java:
```diff
@@ -92,7 +102,7 @@ public String getFormattedUserData() {
             return "";
         }
-        return getUsername() + ":" + this.getPassword();
+        return getUsername().get() + ":" + this.getPassword().get();
```
What's the purpose of this? Isn't the password supposed to be almost always absent in the oauth branch?

irenenikk replied:
It's true that password is actually always `null` in this branch. However to me it is unclear where the method is used in general (it is a part of the Settings interface) , so I thought that the point of the method is to get the password as it returned it before.

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java:
```diff
@@ -116,8 +126,58 @@ public Path getConfigRoot() {
-    public void setCourse(Course course) {}
+    public String hostProgramName() {
+        // which command line is used
+        return null;
```
I think this should be implemented.

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java:
```diff
@@ -116,8 +126,58 @@ public Path getConfigRoot() {
+    @Override
+    public String hostProgramVersion() {
+        // which command line is used
+        return null;
```
You probably want to implement this too

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java:
```diff
@@ -116,8 +126,58 @@ public Path getConfigRoot() {
     @Override
-    public void setConfigRoot(Path path) {}
+    public boolean getSendDiagnostics() {
+        return false;
```
This should be an user-controlled setting. If it's always false, we won't ever get crash reports.

irenenikk replied:
There is no setter for this method, so how is the behaviour determined?

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java:
```diff
@@ -116,8 +126,58 @@ public Path getConfigRoot() {
+    }
+
+    @Override
+    public void setConfigRoot(Path path) {
```
What is `setConfigRoot`? :confused:

irenenikk replied:
This has been discussed earlier and apparently one shouldn't be able to change the config root, so the method is pointless (NOP in tmc netbeans as well). It should be removed from the API.

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/SettingsIo.java:
```diff
@@ -142,7 +142,6 @@ private static boolean requireConfigDirectory(Path configRoot) {
             try {
                 Files.createDirectories(configRoot);
             } catch (Exception e) {
-                //TODO print error to user
```
Why did you remove this TODO? I don't think logger.error prints a message to the user or does it?

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/SettingsIo.java:
```diff
@@ -156,7 +155,6 @@ private static AccountList getHolderFromJson(Path file) {
         try {
             reader = Files.newBufferedReader(file, Charset.forName("UTF-8"));
         } catch (IOException e) {
-            //TODO print error to user
```
See above.

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/SettingsIo.java:
```diff
@@ -169,7 +167,6 @@ private static boolean saveHolderToJson(AccountList holder, Path file) {
         try {
             Files.write(file, json);
         } catch (IOException e) {
-            //TODO print error to user
```
See the above of above :stuck_out_tongue:

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/TmcUtil.java:
```diff
@@ -216,34 +233,46 @@ private static void handleTmcExceptions(CliContext ctx, Exception exception) {
         logger.error("Command failed in tmc-core", exception);
         //TODO we seem to write twice error message; here and in the commands.
-        io.errorln("Command failed in tmc-core, check tmc-cli.log file for more info");
+        io.errorln("Command failed, check tmc-cli.log file for more info");
```
Yes :+1:. The tmc-core part was really confusing to the user.

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/TmcUtil.java:
```diff
@@ -216,34 +233,46 @@ private static void handleTmcExceptions(CliContext ctx, Exception exception) {
+        io.errorln("Command failed, check tmc-cli.log file for more info");
     }
 
     private static boolean isAuthenticationError(Exception exception) {
```
Could be simplified to `return exception.getCause() instanceof OAuthProblemException`.

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/utils/OptionalToGoptional.java:
```diff
@@ -0,0 +1,13 @@
+
+import com.google.common.base.Optional;
+
+public class OptionalToGoptional {
```
:camel:

nygrenh commented on src/test/java/fi/helsinki/cs/tmc/cli/backend/CourseInfoIoTest.java:
```diff
@@ -53,4 +57,15 @@ public void loadingFromFileWorks() {
+        Assert.assertNotNull(loadedInfo);
+
+        CourseInfoIo.abortCreatingCourse(new Course("test-course"), Paths.get(tempDir));
+        File courseJson = Paths.get(tempDir).resolve(".tmc.json").toFile();
```
Avoid using `java.io.File` in new code and prefer the better alternatives in `java.nio`.

For example, this could be rewritten as:
```java
Path courseJson = Paths.get(tempDir).resolve(".tmc.json");
assertTrue(Files.exist(courseJson));
```

irenenikk replied:
:whale:

nygrenh commented on src/test/java/fi/helsinki/cs/tmc/cli/io/ExternalUtilTest.java:
```diff
@@ -23,7 +9,7 @@
 import org.powermock.modules.junit4.PowerMockRunner;
 
-import java.awt.Desktop;
+import java.awt.*;
```
Avoid star imports

irenenikk replied:
It seems like IntelliJ offers star imports, I will stop using the Alt+Enter auto import functionality.

nygrenh commented on src/test/java/fi/helsinki/cs/tmc/cli/io/ExternalUtilTest.java:
```diff
@@ -33,6 +19,16 @@
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.*;
```
Avoid star imports

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/command/OrganizationCommand.java:
```diff
@@ -0,0 +1,118 @@
+                        others.add(o);
+                    }
+                });
+        System.out.println();
```
Some parts of the code use `System.println` for printing and other parts `io.println`.

Commit:
Remove server prompt and useless Settings methods

Commit:
Add functionality to config command

Enable changing server address
Enable toggling sending diagnostics
Only allow certain keys
	- server address
	- send diagnostics
	- update date
	- progress/test bar colors

Commit:
Fix saving configurations to accountList on changes

Commit:
Fix organization command if not logged in

Also small tweaks to prints and tests

Commit:
Remove java jdk from travis.yml

Lolled

Commit:
Update TMC Core version

Commit:
Update depedencies

Commit:
Add setCourse for Settings

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/CourseInfoIo.java:
```diff
@@ -62,4 +62,17 @@ public static void createNewCourse(Course course, Account account, Path parentDi
         CourseInfoIo.save(info, configFile);
     }
+
+    public static void abortCreatingCourse(Course course, Path parentDir) {
```
Bit misleading name

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java:
```diff
@@ -49,32 +48,37 @@ public void setWorkDir(WorkDir workDir) {
     @Override
     public String getServerAddress() {
-        return account.getServerAddress();
+            return account.getServerAddress();
```
Why did you add this indentation?

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/backend/SettingsIo.java:
```diff
@@ -67,6 +70,24 @@ public static boolean delete() {
+    public static void saveCurrentSettingsToAccountList(Settings settings) {
+        AccountList list = loadAccountList();
+        Set<Account> deletables = new HashSet<>();
+        for (Iterator<Account> i = list.iterator(); i.hasNext(); ) {
```
For each?

nygrenh commented on src/main/java/fi/helsinki/cs/tmc/cli/shared/CourseFinder.java:
```diff
@@ -88,7 +88,7 @@ private boolean handleMultipleMatchingCourses(Map<Account, Course> matches) {
                             + entryAccount.getServerAddress()
                             + " with '"
-                            + entryAccount.getUsername()
+                            + entryAccount.getUsername().get()
```
Nullpointerexception?

Commit:
Make small refactorings
